### PR TITLE
Add TRTLLM MoE NVFP4 kernel to CompressedTensorsW4A4MoeMethod

### PR DESCRIPTION
**Pull request by Victor49152**
## Purpose
This PR is created for two purposes:
1. Add trtllm fp4 kernel to be used in compressedtensorw4a4 method
2. Create a mutually used function for modelopt and compressedtensorw4a4 path to call flashinfer trtllm moe fp4 kernel

### Accuracy test
Tested all code paths before and after the code change, including:
1. modelopt + trtllm fp4 moe
2. modelopt + flashinfer cutlass moe
3. compressedtensor + cutlass moe
4. compressedtensor + flashinfer cutlass moe
5. compressedtensor + flashinfer trtllm fp4 moe
The eval task is gsm8k and results are listed below 

The eval is based on nvfp4 ckpoint generated by llm-compressor for Qwen3-VL-235B-A22B for compressed tensors path and nvidia/Qwen3-235B-A22B-FP4 on modelopt path. The serving command is 

`vllm serve model_name    --tensor-parallel-size 2   --mm-encoder-tp-mode data --enable-expert-parallel --async-scheduling&
`
and eval command is 
`lm_eval --model local-completions --tasks gsm8k --model_args model=model_name,base_url=http://0.0.0.0:8000/v1/completions,max_retries=3,tokenized_requests=False,timeout=1200,max_gen_toks=2048,max_length=8192 --batch_size 2048 --trust_remote_code --limit 0.5`

### Perf test

Perf test is done at single GB200 and TEP=2 at GB200

The checkpoint used is RedHatAI/Qwen3-VL-235B-A22B-Instruct-NVFP4, server start command is 
`vllm serve RedHatAI/Qwen3-VL-235B-A22B-Instruct-NVFP4 --host 0.0.0.0 --port 46503 -tp <tp_size> -dp 1 --mm-encoder-tp-mode data --enable-expert-parallel --async-scheduling --max-model-len 32768 --cudagraph-capture-sizes 1 2 4 8 16 24 32 40 48 56 64 72 80 88 96 104 112 120 128 136 144 152 160 168 176 184 192 200 208 216 224 232 240 248 256 272 288 304 320 336 352 368 384 400 416 432 448 464 480 496 512 1024 2048 4096 8192 --max-num-seqs 1024`

The eval command is 
`vllm bench serve   --backend openai-chat   --endpoint /v1/chat/completions   --model RedHatAI/Qwen3-VL-235B-A22B-Instruct-NVFP4   --dataset-name random   --random-input-len 300   --random-output-len 200   --random-mm-base-items-per-request 1   --random-mm-num-mm-items-range-ratio 0   --num-prompts 1250     --random-mm-limit-mm-per-prompt 1   --random-mm-bucket-config '{(512, 512, 1): 1.0}'   --ignore-eos   --base-url http://0.0.0.0:46503
`

### Accuracy test result

1. modelopt + trtllm fp4 moe

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9348|±  |0.0096|
|     |       |strict-match    |     5|exact_match|↑  |0.9303|±  |0.0099|

2. modelopt + flashinfer cutlass moe

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9348|±  |0.0096|
|     |       |strict-match    |     5|exact_match|↑  |0.9333|±  |0.0097|

3. compressedtensor + cutlass moe

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9606|±  |0.0076|
|     |       |strict-match    |     5|exact_match|↑  |0.9561|±  |0.0080|

4. compressedtensor + flashinfer cutlass moe

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9545|±  |0.0081|
|     |       |strict-match    |     5|exact_match|↑  |0.9409|±  |0.0092|

5. compressedtensor + flashinfer trtllm fp4 moe

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9606|±  |0.0076|
|     |       |strict-match    |     5|exact_match|↑  |0.9485|±  |0.0086|

#### TEP=2 on GB200, trtllm kernel is 81.3% faster than flashinfer.cutlass in request throughput

Single GPU flashinfer.cutlass kernel 

| Metric                                   | Value       |
|-------------------------------------------|-------------|
| **Successful requests**                   | 1250        |
| **Failed requests**                       | 0           |
| **Benchmark duration (s)**                | 44.64       |
| **Total input tokens**                    | 375000      |
| **Total generated tokens**                | 250000      |
| **Request throughput (req/s)**            | 28.00       |
| **Output token throughput (tok/s)**       | 5599.97     |
| **Peak output token throughput (tok/s)**  | 10613.00    |
| **Peak concurrent requests**              | 1250.00     |
| **Total Token throughput (tok/s)**        | 13999.93    |
| **Mean TTFT (ms)**                        | 16529.75    |
| **Median TTFT (ms)**                      | 16429.34    |
| **P99 TTFT (ms)**                         | 38471.74    |
| **Mean TPOT (ms)**                        | 57.15       |
| **Median TPOT (ms)**                      | 55.44       |
| **P99 TPOT (ms)**                         | 88.73       |
| **Mean ITL (ms)**                         | 56.87       |
| **Median ITL (ms)**                       | 41.37       |
| **P99 ITL (ms)**                           | 156.58      |

Single GPU flashinfer.TRTLLM kernel

| Metric                                   | Value       |
|-------------------------------------------|-------------|
| **Successful requests**                   | 1250        |
| **Failed requests**                       | 0           |
| **Benchmark duration (s)**                | 36.14       |
| **Total input tokens**                    | 375000      |
| **Total generated tokens**                | 250000      |
| **Request throughput (req/s)**            | 34.58       |
| **Output token throughput (tok/s)**       | 6916.91     |
| **Peak output token throughput (tok/s)**  | 11581.00    |
| **Peak concurrent requests**              | 1250.00     |
| **Total Token throughput (tok/s)**        | 17292.27    |
| **Mean TTFT (ms)**                        | 13332.46    |
| **Median TTFT (ms)**                      | 13195.51    |
| **P99 TTFT (ms)**                         | 31091.85    |
| **Mean TPOT (ms)**                        | 46.46       |
| **Median TPOT (ms)**                      | 45.06       |
| **P99 TPOT (ms)**                         | 72.82       |
| **Mean ITL (ms)**                         | 46.24       |
| **Median ITL (ms)**                       | 36.34       |
| **P99 ITL (ms)**                          | 105.93      |

TEP=2 flashinfer.cutlass kernel 

| Metric                                   | Value       |
|-------------------------------------------|-------------|
| **Successful requests**                   | 1250        |
| **Failed requests**                       | 0           |
| **Benchmark duration (s)**                | 25.16       |
| **Total input tokens**                    | 375000      |
| **Total generated tokens**                | 250000      |
| **Request throughput (req/s)**            | 49.68       |
| **Output token throughput (tok/s)**       | 9935.25     |
| **Peak output token throughput (tok/s)**  | 15659.00    |
| **Peak concurrent requests**              | 1250.00     |
| **Total Token throughput (tok/s)**        | 24838.14    |
| **Mean TTFT (ms)**                        | 8506.92     |
| **Median TTFT (ms)**                      | 7748.88     |
| **P99 TTFT (ms)**                         | 18321.49    |
| **Mean TPOT (ms)**                        | 57.87       |
| **Median TPOT (ms)**                      | 63.02       |
| **P99 TPOT (ms)**                         | 66.68       |
| **Mean ITL (ms)**                         | 57.59       |
| **Median ITL (ms)**                       | 63.96       |
| **P99 ITL (ms)**                          | 78.77       |

TEP=2 flashinfer.trtllm kernel 

| Metric                                   | Value       |
|-------------------------------------------|-------------|
| **Successful requests**                   | 1250        |
| **Failed requests**                       | 0           |
| **Benchmark duration (s)**                | 13.88       |
| **Total input tokens**                    | 375000      |
| **Total generated tokens**                | 250000      |
| **Request throughput (req/s)**            | 90.07       |
| **Output token throughput (tok/s)**       | 18013.18    |
| **Peak output token throughput (tok/s)**  | 24576.00    |
| **Peak concurrent requests**              | 1250.00     |
| **Total Token throughput (tok/s)**        | 45032.94    |
| **Mean TTFT (ms)**                        | 2727.64     |
| **Median TTFT (ms)**                      | 1284.30     |
| **P99 TTFT (ms)**                         | 10048.45    |
| **Mean TPOT (ms)**                        | 39.57       |
| **Median TPOT (ms)**                      | 43.89       |
| **P99 TPOT (ms)**                         | 44.84       |
| **Mean ITL (ms)**                         | 39.51       |
| **Median ITL (ms)**                       | 42.41       |
| **P99 ITL (ms)**                          | 91.95       |

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by mgoin** [comment]
@Victor49152 sorry for the possible thrash, I was resolving the lint and merge conflicts

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 1]
```diff
@@ -369,22 +498,92 @@ def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
...
+            self.allow_flashinfer
+            and self.flashinfer_moe_backend == FlashinferMoeBackend.TENSORRT_LLM
+        ):
+            w2_input_global_scale = (
+                layer.w2_input_global_scale.min()
+                .to(torch.float32)
+                .expand(layer.num_experts)
+            )
+        else:
+            w2_input_global_scale = layer.w2_input_global_scale
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:395]: Why don't we need to do the same processing for w13_input_global_scale? I see that modelopt seems to do both

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py [hunk 2]
```diff
@@ -304,6 +318,130 @@ def create_weights(
         )
         set_weight_attrs(w2_input_scale, extra_weight_attrs)
 
+    def prepare_static_weights_for_trtllm_fp4_moe(
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: nit: we should generalize this as well between ct and modelopt

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 1 above)
Comment (Victor49152) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:395]: From what I inspect from the both quantized checkpoint for Qwen, this w13 input global scale contains identical values for all elements, and the kernel would take the first element anyway. So I didn't modify it to keep the change minimum. @jiahanc Do you know if that is the case always for all models? If not we should add the min() to w13_input_global_scale, too

Comment (jiahanc) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:395]: Not sure about compressed tensor, but in modelopt checkpoint, the values can vary in `w13_input_global_scale`, so we need to do the max() to get the global SF

Comment (Victor49152) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:395]: I see, I will add the change and retest the results shortly. Thanks

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 2 above)
Comment (Victor49152) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: SGTM, on it

Comment (Victor49152) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py]: It's moved to utils too

File: vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py (hunk 1 above)
Comment (Victor49152) [reply vllm/model_executor/layers/quantization/compressed_tensors/compressed_tensors_moe.py:395]: I added the same processing to w13_input_global_scale, and re-run all the accuracy check on 235B models, results are updated in the PR description/

**Review by mgoin — APPROVED** [review]
Great work @Victor49152 ! I appreciate the careful testing and reporting
